### PR TITLE
feat(positioning): add positioning service

### PR DESCRIPTION
**Pull request by RobJacobs**
Closes #289

Need feedback/help with::

1.  Is the name 'Positioning' acceptable?
2.  Does the Positioning class need to be marked as @Injectable?
3.  I'm getting conflicting clang-format and tslint errors on spaces around curly braces, tslint wants:

` func(): number { return 0; }`

clang-format wants:

` func(): number{return 0;}`

to get the build to complete locally, I had to remove the tslint rules.
4.  Rather than create an ELementPosition interface, should I use the ClientRect interface instead?  It has 2 properties (right and bottom) that we don't currently return, but we could easily enough. 

I still need to write the test, but I wanted to make sure I'm going down the right path, thanks.

### CODE REVIEW

File: src/util/positioning.ts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    let parentOffset: ElementPosition = {width: 0, height: 0, top: 0, left: 0};
+    let offsetParentEl: HTMLElement = this.parentOffsetElement(element);
+
+    if (offsetParentEl !== document.documentElement) {
+      parentOffset = this.offset(offsetParentEl);
+      parentOffset.top += offsetParentEl.clientTop - offsetParentEl.scrollTop;
+      parentOffset.left += offsetParentEl.clientLeft - offsetParentEl.scrollLeft;
+    }
+
+    return <ElementPosition>{
```
Comment (icfantv) [inline src/util/positioning.ts]: don't need the cast here.

File: src/util/positioning.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      width: elBCR.width || element.offsetWidth,
+      height: elBCR.height || element.offsetHeight,
+      top: elOffset.top - parentOffset.top,
+      left: elOffset.left - parentOffset.left
+    };
+  }
+
+  offset(element: HTMLElement): ElementPosition {
+    let elBcr: ClientRect = element.getBoundingClientRect();
+    return <ElementPosition>{
```
Comment (icfantv) [inline src/util/positioning.ts]: ditto

File: src/util/positioning.ts (hunk 1 above)
Comment (RobJacobs) [inline src/util/positioning.ts]: It adds type safety to the object I'm returning, or is that not important?

Comment (icfantv) [inline src/util/positioning.ts]: the compiler will check that if the function declares a return type, that the returned type matches what's declared.

File: src/util/positioning.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
+import {ElementRef} from '@angular/core';
+
+export interface ElementPosition {
```
Comment (icfantv) [inline src/util/positioning.ts]: should be named `IElementPosition` since it is an interface

File: src/util/positioning.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
+import {ElementRef} from '@angular/core';
```
Comment (pkozlowski-opensource) [inline src/util/positioning.ts]: Doesn't look like we are using `ElementRef` anywhere (and probably shouldn't be) so this import can be removed.

**Comment by pkozlowski-opensource** [comment]
@RobJacobs I will review in grater details tomorrow but I was wondering if there is any way to write meaningful tests for this service?

**Comment by RobJacobs** [comment]
@pkozlowski-opensource I'm going to look at how material is testing their stuff and see what we can do.

File: src/util/positioning.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      offsetParentEl = offsetParentEl.offsetParent as HTMLElement;
+    }
+    return offsetParentEl || document.documentElement;
+  }
+
+  position(element: HTMLElement): ElementPosition {
+    let elOffset: ElementPosition = this.offset(element);
+    let elBCR: ClientRect = element.getBoundingClientRect();
+    let parentOffset: ElementPosition = {width: 0, height: 0, top: 0, left: 0};
+    let offsetParentEl: HTMLElement = this.parentOffsetElement(element);
```
Comment (wesleycho) [inline src/util/positioning.ts]: The typings in these several of lines are unnecessary I think

Comment (wesleycho) [inline src/util/positioning.ts]: Well, excepting `parentOffset` - that is probably ok.

File: src/util/positioning.ts [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    };
+  }
+
+  positionElements(hostElement: HTMLElement, targetElement: HTMLElement, placement: string, appendToBody: boolean):
+      ElementPosition {
+    let hostElPosition: ElementPosition = appendToBody ? this.offset(hostElement) : this.position(hostElement);
+    let shiftWidth: any = {
+      left(): number{return hostElPosition.left;},
+      center(): number{return hostElPosition.left + hostElPosition.width / 2 - targetElement.offsetWidth / 2;},
+      right(): number{return hostElPosition.left + hostElPosition.width;}
```
Comment (wesleycho) [inline src/util/positioning.ts]: How come these have to be functions? Is this to readjust after the element has been rendered?

File: src/util/positioning.ts [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    };
+    let shiftHeight: any = {
+      top(): number{return hostElPosition.top;},
+      center(): number{return hostElPosition.top + hostElPosition.height / 2 - targetElement.offsetHeight / 2;},
+      bottom(): number{return hostElPosition.top + hostElPosition.height;}
```
Comment (wesleycho) [inline src/util/positioning.ts]: Same question as above

File: src/util/positioning.ts [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    };
+    let targetElBCR: ClientRect = targetElement.getBoundingClientRect();
```
Comment (wesleycho) [inline src/util/positioning.ts]: Typing can probably be removed as inferred.

**Comment by pkozlowski-opensource** [comment]
@RobJacobs as a very minimum we should have tests for calculations (those could be decoupled from the DOM - that is we could write tests for the part that does calculations after dimensions of all the involved elements is known).

**Comment by RobJacobs** [comment]
Updated based on suggestions.  I switched from creating an ElementPosition interface to using the Angular ClientRect interface and now return right/bottom values.  That will put us in a better position to support smart positioning down the road.  I'm going to work on tests today.

**Comment by RobJacobs** [comment]
@pkozlowski-opensource Tests added, did a little more refactoring.  I think this is good to go, let me know if you see anything that needs work.

File: src/util/positioning.spec.ts [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+  const bodyHeight = document.body.style.height;
+  const bodyWidth = document.body.style.width;
+
+  function createElement(height: number, width: number, marginTop: number, marginLeft: number): HTMLElement {
+    let element = document.createElement('div');
+    element.style.display = 'inline-block';
+    element.style.height = height + 'px';
+    element.style.width = width + 'px';
+    element.style.marginTop = marginTop + 'px';
+    element.style.marginLeft = marginLeft + 'px';
```
Comment (wesleycho) [inline src/util/positioning.spec.ts]: Could use template strings for these 4 lines

**Comment by pkozlowski-opensource** [comment]
@RobJacobs thnx, I've merged it to unblock the situation. We know that we need to improve it but I think it is good as a starting point. Expect to hear from me more as soon as I start impl of tooltips / popovers (just after alpha.0 is out).

**Pull request closed without merging**